File: Driver Web App/src/services/firebaseService.js
/**
 * NTFMS Driver Portal — Firestore Service Layer
 *
 * Replaces localStorage calls with Firestore operations.
 * The Admin Web Portal and Mobile App use the same collections,
 * so all three apps share live data automatically.
 *
 * Collections:
 *   fines       — Traffic fine records (doc ID = refNo)
 *   categories  — Fine category definitions (doc ID = category ID)
 */

import {
  collection,
  doc,
  getDoc,
  getDocs,
  updateDoc,
  Timestamp,
  query,
  where,
} from 'firebase/firestore';
import { db } from '../firebase';

const finesCol      = collection(db, 'fines');
const categoriesCol = collection(db, 'categories');

// ── Helpers ───────────────────────────────────────────────────────

/**
 * Convert a Firestore Timestamp (or ISO string) to a JS Date ISO string.
 * Falls back to null gracefully.
 */
function normaliseDate(val) {
  if (!val) return null;
  if (val instanceof Timestamp) return val.toDate().toISOString();
  if (typeof val === 'string') return val;
  return null;
}

/**
 * Normalise a raw Firestore fine document so downstream components
 * receive plain JS objects with ISO date strings (same shape as before).
 */
function normaliseFine(data) {
  return {
    ...data,
    issuedAt: normaliseDate(data.issuedAt),
    paidAt:   normaliseDate(data.paidAt),
  };
}

// ── Public API ────────────────────────────────────────────────────

/**
 * Look up a fine by reference number AND category ID.
 * Returns the fine document enriched with category name & penalty points.
 * Throws a structured error if not found.
 */
export async function lookupFine(referenceNumber, categoryId) {
  const refUpper = referenceNumber.trim().toUpperCase();
  const catUpper = categoryId.trim().toUpperCase();

  // Fine doc ID = refNo (set during seeding)
  const fineRef  = doc(finesCol, refUpper);
  const fineSnap = await getDoc(fineRef);

  if (!fineSnap.exists()) {
    throw {
      code: 404,
      message:
        'Fine not found. Please verify the Reference Number and Category ID from your fine sheet and try again.',
    };
  }

  const fineData = fineSnap.data();

  // Verify category matches
  if (fineData.category.toUpperCase() !== catUpper) {
    throw {
      code: 404,
      message:
        'Fine not found. Please verify the Reference Number and Category ID from your fine sheet and try again.',
    };
  }

  // Fetch category details for name + penaltyPoints
  const catRef  = doc(categoriesCol, fineData.category);
  const catSnap = await getDoc(catRef);
  const catData = catSnap.exists() ? catSnap.data() : {};

  return {
    ...normaliseFine(fineData),
    categoryName:  catData.name          ?? 'Unknown Category',
    penaltyPoints: catData.penaltyPoints ?? 0,
  };
}

/**
 * Mark a fine as Paid in Firestore.
 * Called after card validation succeeds in api.js.
 *
 * @param {string} refNo  — Fine reference number (e.g. "SLP-2026-9814")
 * @param {string} method — Payment method label ("Web Portal")
 * @returns {object} Updated fine data
 */
export async function markFinePaid(refNo, method = 'Web Portal') {
  const fineRef  = doc(finesCol, refNo);
  const fineSnap = await getDoc(fineRef);

  if (!fineSnap.exists()) {
    throw { code: 404, message: `Fine reference ${refNo} not found in the national register.` };
  }

  const fineData = fineSnap.data();

  if (fineData.status === 'Paid') {
    throw { code: 409, message: `Fine ${refNo} has already been settled. No duplicate payment is required.` };
  }

  const paidAt = Timestamp.now();

  await updateDoc(fineRef, {
    status:        'Paid',
    paymentMethod: method,
    paidAt,
    smsSent:       true,
  });

  return normaliseFine({
    ...fineData,
    status:        'Paid',
    paymentMethod: method,
    paidAt,
    smsSent:       true,
  });
}

/**
 * Fetch all fine categories.
 * @returns {Array} Array of category objects
 */
export async function getCategories() {
  const snap = await getDocs(categoriesCol);
  return snap.docs.map(d => ({ id: d.id, ...d.data() }));
}

/**
 * Fetch all fines associated with a driver's license number.
 */
export async function getFinesByLicense(licenseNumber) {
  if (!licenseNumber || licenseNumber.trim() === '') return [];
  const q = query(finesCol, where('driverLicense', '==', licenseNumber.trim().toUpperCase()));
  const snap = await getDocs(q);
  const list = snap.docs.map(d => {
    return {
      refNo: d.id,
      ...normaliseFine(d.data())
    };
  });
  list.sort((a, b) => new Date(b.issuedAt) - new Date(a.issuedAt));
  return list;
}

File: admin-portal/MobileApp/lib/services/firebase_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/fine_model.dart';

/// NTFMS Mobile — Firestore Service Layer
///
/// Reads from and writes to the shared Firestore database (project: ntfms2026).
/// The same database is used by the Admin Web Portal and Driver Web Portal,
/// so all three apps see live, consistent data.
///
/// Collections:
///   fines       — Traffic fine records   (doc ID = refNo, e.g. "SLP-2026-9814")
///   categories  — Fine category metadata (doc ID = "CAT-01" … "CAT-09")
class FirebaseService {
  static final FirebaseFirestore _db = FirebaseFirestore.instance;
  static final CollectionReference _finesCol      = _db.collection('fines');
  static final CollectionReference _categoriesCol = _db.collection('categories');

  // ── Helpers ─────────────────────────────────────────────────────

  /// Convert a Firestore Timestamp or ISO string to DateTime.
  static DateTime _parseDate(dynamic val) {
    if (val == null) return DateTime.now();
    if (val is Timestamp) return val.toDate();
    if (val is String) {
      try { return DateTime.parse(val); } catch (_) { return DateTime.now(); }
    }
    return DateTime.now();
  }

  /// Convert a Firestore document snapshot to a FineModel.
  static FineModel _docToFine(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return FineModel(
      fineReferenceNumber:  doc.id,
      categoryId:           data['category']      as String? ?? '',
      fineAmount:           (data['amount']        as num?)?.toDouble() ?? 0.0,
      violatorName:         data['driverName']     as String? ?? '',
      violatorLicenseNumber:data['driverLicense']  as String? ?? '',
      violationType:        data['category']       as String? ?? '',
      issuedDate:           _parseDate(data['issuedAt']),
      status:               data['status']         as String? ?? 'Pending',
      locationDescription:  data['location']       as String?,
      officerContactNumber: data['officerPhone']   as String?,
      vehicleNo:            data['vehicleNo']      as String? ?? '',
    );
  }

  // ── Public API ───────────────────────────────────────────────────

  /// Fetch fine details by reference number AND category ID.
  ///
  /// Uses the document ID (= refNo) for an O(1) Firestore read.
  /// Throws if the fine doesn't exist or the category doesn't match.
  static Future<FineModel> fetchFineDetails({
    required String fineReferenceNumber,
    required String categoryId,
  }) async {
    if (fineReferenceNumber.trim().isEmpty || categoryId.trim().isEmpty) {
      throw Exception('Fine Reference Number and Category ID are required');
    }

    final refUpper = fineReferenceNumber.trim().toUpperCase();
    final catUpper = categoryId.trim().toUpperCase();

    final snap = await _finesCol.doc(refUpper).get();

    if (!snap.exists) {
      throw Exception(
        'Fine not found. Please verify the Reference Number and Category ID.',
      );
    }

    final data = snap.data() as Map<String, dynamic>;
    final storedCategory = (data['category'] as String? ?? '').toUpperCase();

    if (storedCategory != catUpper) {
      throw Exception(
        'Fine not found. Please verify the Reference Number and Category ID.',
      );
    }

    return _docToFine(snap);
  }

  /// Mark a fine as Paid in Firestore and return the updated model.
  ///
  /// Called after client-side card validation succeeds.
  static Future<Map<String, dynamic>> processPayment({
    required String fineReferenceNumber,
    required String categoryId,
    required Map<String, dynamic> paymentDetails,
  }) async {
    if (fineReferenceNumber.trim().isEmpty ||
        categoryId.trim().isEmpty ||
        paymentDetails.isEmpty) {
      throw Exception('Missing required payment information');
    }

    final refUpper = fineReferenceNumber.trim().toUpperCase();
    final docRef   = _finesCol.doc(refUpper);
    final snap     = await docRef.get();

    if (!snap.exists) {
      throw Exception('Fine reference $refUpper not found in the national register.');
    }

    final data   = snap.data() as Map<String, dynamic>;
    final status = data['status'] as String? ?? '';

    if (status == 'Paid') {
      throw Exception('This fine has already been paid.');
    }

    // Mark as Paid in Firestore
    final paidAt = Timestamp.now();
    await docRef.update({
      'status':        'Paid',
      'paymentMethod': paymentDetails['paymentMethod'] ?? 'Mobile App',
      'paidAt':        paidAt,
      'smsSent':       true,
    });

    // Generate confirmation identifiers
    final confirmationNumber = 'CONF-${DateTime.now().year}-${(10000 + DateTime.now().millisecondsSinceEpoch % 90000).toInt()}';
    final transactionId      = 'TXN-${(100000 + DateTime.now().millisecondsSinceEpoch % 900000).toInt()}';

    return {
      'success':            true,
      'message':            'Payment processed successfully',
      'confirmationNumber': confirmationNumber,
      'transactionId':      transactionId,
      'smsStatus':          'sent',
    };
  }

  /// Fetch all fine categories from Firestore.
  static Future<List<Map<String, dynamic>>> getCategories() async {
    final snap = await _categoriesCol.get();
    return snap.docs.map((d) {
      final data = d.data() as Map<String, dynamic>;
      return { 'id': d.id, ...data };
    }).toList();
  }
}
